STM32 ADC minor cleanup and rework

gautierg-st opened the pull request:
This PR cleans up a bit the code of the STM32 ADC driver:
- Remove some redundant code
- Invert the logic of some #ifdef, to generally reduce length, and make integration of future series easier
- Minor rework of calibration function for F1 and F37x
- Add some missing ifdef for F37x
- Add and fix some comments

gautierg-st committed:
drivers: adc: rewrite stm32 adc enable

Some STM32 families that need to check the ADRDY flag after enabling the
ADC were not doing it. The #ifdef has been updated to fix that.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: clean stm32 adc calibration

Rework and clean some code around STM32 ADC calibration

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: add comment to describe different adc versions

Add comment to describe different ADC versions

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: cleanup stm32 adc driver

Inverts the logic of some #ifdef, replacing lists of newer series by a
list of NOT older series. This makes it shorter, and more future-proof as
future series are more likely to work out of the box, without need to
manually add a new define in each place.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: add ifdef for stm32f37x adc

Add ifdef around some functions that don't exist for STM32F37x (ADC_V2_5)
like it is done for STM32F1x.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -648,27 +648,25 @@ static int adc_stm32_enable(ADC_TypeDef *adc)
 	uint32_t count_timeout = 0;
 
 	while (LL_ADC_IsActiveFlag_ADRDY(adc) == 0) {
+		/* For F0, continue to write ADEN=1 until ADRDY=1 */
 		if (LL_ADC_IsEnabled(adc) == 0UL) {
 			LL_ADC_Enable(adc);
 		}
```
Should it be F0 specific then ?

gautierg-st replied:
Possibly. I found this note only in F0 reference manual. It is likely not necessary on other series. I'll try making it F0 specific and launching some tests

gautierg-st replied:
L1 is also different from F2/F4/F7 regarding the ADC prescaler. But that is a whole separate topic on its own.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1089,30 +1090,21 @@ static int start_read(const struct device *dev,
-	defined(CONFIG_SOC_SERIES_STM32U5X) || \
-	defined(CONFIG_SOC_SERIES_STM32WLX)
-	LL_ADC_EnableIT_EOC(adc);
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X) || \
+	defined(CONFIG_SOC_SERIES_STM32L1X)
```
Looks like we can start thinking about introducing IP dedicated compatibles with for instance st,stm32f4-adc which could be used by F2/F4/F7/L1 series. Am I right ?

gautierg-st replied:
Yes, every time these 4 are defined together. This was also the case for the resolution. Sampling time values are different for L1, but that is treated in PR #57342

erwango commented:
@xyzzy42, @heinwessels  you might be interested

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -232,6 +240,12 @@ static const uint32_t table_samp_time[] = {
+ * Number of ADC clock cycles to wait before starting calibration and after
+ * enabling ADC for STM32F1 and STM32F37x
+ */
+#define STM32F1_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES	2
```
this is: 
#define STM32F1_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES  LL_ADC_DELAY_ENABLE_CALIB_ADC_CYCLES

gautierg-st replied:
Correct, I didn't see that. There is yet another macro for STM32F37x. I'll update this part

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -1526,15 +1480,16 @@ static int adc_stm32_init(const struct device *dev)
-	LL_ADC_StartCalibration(adc);
+#if defined(CONFIG_SOC_SERIES_STM32F1X) || defined(STM32F3X_ADC_V2_5)
+	wait_cycles = SystemCoreClock / adc_rate *
+		      STM32F1_ADC_DELAY_CALIB_ENABLE_ADC_CYCLES;
```
could adc_rate  be null ?

gautierg-st replied:
It clearly shouldn't be, but in doubt, I added a check. Also moved this waiting in a specific function.

FRASTM reviewed (commented):
LGTM

gautierg-st committed:
dts: bindings: adc: add two compat for stm32 adc

Add two compat for STM32F4-like (F2, F4, F7 & L1) and STM32F1-like (F1
& F37x) ADCs.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: add new adc compatible to stm32

Add the new ADC compatibles for STM32F1 & F373, and for F2, F4, F7 & L1.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: make use of new stm32 adc compatibles

Several sections of the STM32 ADC driver are #ifdef by a combination of the
same SOC defines that share similar IPs. These are F2/F4/F7/L1, and
F1/F37x.
Each of these combinations is now replaced by a specific compatible, which
makes the code a bit lighter and more succinct.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

trentp-igor commented on dts/arm/st/f1/stm32f1.dtsi:
```diff
@@ -333,7 +333,7 @@
 
 		adc1: adc@40012400 {
-			compatible = "st,stm32-adc";
+			compatible = "st,stm32-adc", "st,stm32f1-adc";
```
The normal convention is for the most specific compatible to be listed first and then to get more generic.  So stm32f1-adc would be first, then stm32-adc.

gautierg-st replied:
Thanks, I wasn't aware of this convention. Fixed now.

gautierg-st committed:
dts: arm: st: h7: remove adc3 for stm32h7ax

STM32H7Ax/H7Bx have only two ADCs, so we delete the ADC3 node.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>
